application/src/main/java/org/togetherjava/tjbot/logging/discord/DiscordLogForwarder.java

```diff
@@ -0,0 +1,236 @@
+package org.togetherjava.tjbot.logging.discord;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.logging.log4j.core.LogEvent;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.logging.LogMarkers;
+import org.togetherjava.tjbot.logging.discord.api.DiscordLogBatch;
+import org.togetherjava.tjbot.logging.discord.api.DiscordLogMessageEmbed;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Instant;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+/**
+ * Forwards log events to a Discord channel via a webhook. See {@link #forwardLogEvent(LogEvent)}.
+ * <p>
+ * Logs are forwarded in correct order, based on their timestamp. They are not forwarded
+ * immediately, but at a fixed schedule in batches of {@value MAX_BATCH_SIZE} logs.
+ * <p>
+ * Failed logs are repeated {@value MAX_RETRIES_UNTIL_DISCARD} times until eventually discarded.
+ * <p>
+ * Although unlikely to hit, the class maximally buffers {@value MAX_PENDING_LOGS} logs until
+ * discarding further logs. Under normal circumstances, the class can easily handle high loads of
+ * logs.
+ */
+final class DiscordLogForwarder {
```

Comment:
This feels like a matter of reinventing the wheel, you're just making a smaller Discord webhooks library, maintaining this can only be done by you, or other expert programmers that want to spend their time into figuring out what you exactrly did.

Why did you not just use an existing library, that does the rate limiting and such for you?

Reply:
> Why did you not just use an existing library, that does the rate limiting and such for you?

Unless you find the wheel I am reinventing here, I am not reinventing anything. I couldnt find anything like that.

But maybe I just looked for the wrong stuff, I was specifically looking for SLF4J2 Log Appenders for Discord. And I only found 2 very unsatisfying implementations (which also did not address this aspect).

If you can link me anything that can reduce the complexity in this code, I would be very glad to replace it.

Reply:
That said, I did my best to make this class maintainable by others. I dont think its _that bad_. It has a few flaws, yes, but it should be doable.

Reply:
It's not that bad, but I still believe that using a library will help.

I'm not talking about a Logback <-> Discord library, buy just a Discord webhook library (https://github.com/MinnDevelopment/discord-webhooks)

This will remove the need of rate limiting, and all the Http stuff.

Reply:
i wasnt aware of such libs. thanks for letting me know. this is a good idea and ill check it out

Reply:
done. recheck please